matlab/src/feather_reader.cc

```diff
@@ -0,0 +1,212 @@
+    case arrow::Type::UINT16:
+      return ReadNumericVariableData<arrow::UInt16Type>(column);
+    case arrow::Type::UINT32:
+      return ReadNumericVariableData<arrow::UInt32Type>(column);
+    case arrow::Type::UINT64:
+      return ReadNumericVariableData<arrow::UInt64Type>(column);
+    case arrow::Type::INT8:
+      return ReadNumericVariableData<arrow::Int8Type>(column);
+    case arrow::Type::INT16:
+      return ReadNumericVariableData<arrow::Int16Type>(column);
+    case arrow::Type::INT32:
+      return ReadNumericVariableData<arrow::Int32Type>(column);
+    case arrow::Type::INT64:
+      return ReadNumericVariableData<arrow::Int64Type>(column);
+
+    default: {
+      mexErrMsgIdAndTxt("MATLAB:arrow:UnsupportedArrowType",
+                        "Unsupported arrow::Type '%s' for variable '%s'",
+                        type->name().c_str(), column->name().c_str());
+      break;
+    }
+  }
+
+  return nullptr;
+}
+
+// Read the nulls of variable i from the Feather file as a mxArray*.
+mxArray* ReadVariableNulls(const std::shared_ptr<arrow::Column>& column) {
+  // TODO: Implement proper null value support. For the time being,
+  // we will simply return a zero initialized logical array to MATLAB.
+  return mxCreateLogicalMatrix(column->length(), 1);
+}
+
+// Read the type of variable i from the Feather file as a mxArray*.
+mxArray* ReadVariableType(const std::shared_ptr<arrow::Column>& column) {
+  return mxCreateString(column->type()->name().c_str());
+}
+
+// MATLAB arrays cannot be larger than 2^48.
+static constexpr uint64_t MAX_MATLAB_SIZE = static_cast<uint64_t>(0x01) << 48;
+
+}  // namespace internal
+
+arrow::Status FeatherReader::Open(const std::string& filename,
+                                  std::shared_ptr<FeatherReader>* feather_reader) {
+  *feather_reader = std::shared_ptr<FeatherReader>(new FeatherReader());
+  // Open file with given filename as a ReadableFile.
+  std::shared_ptr<arrow::io::ReadableFile> readable_file(nullptr);
+  auto status = arrow::io::ReadableFile::Open(filename, &readable_file);
+  if (!status.ok()) {
+    return status;
+  }
+  // TableReader expects a RandomAccessFile.
+  std::shared_ptr<arrow::io::RandomAccessFile> random_access_file(readable_file);
+  // Open the Feather file for reading with a TableReader.
+  status = arrow::ipc::feather::TableReader::Open(random_access_file,
+                                                  &(*feather_reader)->table_reader_);
+  if (!status.ok()) {
+    return status;
+  }
+
+  // Read the table metadata from the Feather file.
+  (*feather_reader)->num_rows_ = (*feather_reader)->table_reader_->num_rows();
+  (*feather_reader)->num_variables_ = (*feather_reader)->table_reader_->num_columns();
+  (*feather_reader)->description_ =
+      (*feather_reader)->table_reader_->HasDescription()
+          ? (*feather_reader)->table_reader_->GetDescription()
+          : "";
+  (*feather_reader)->version_ = (*feather_reader)->table_reader_->version();
```

Comment:
This could be encapsulated in a private initialization method. I don't have a MATLAB installation so I'm not going to attempt to refactor this myself

Reply:
Thanks for this suggestion. I can encapsulate this initialization code as part of a future pull request.